### PR TITLE
UPSTREAM: 21373: kubelet: reading cloudinfo from cadvisor

### PR DESCRIPTION
**Pull request by deads2k**
When no --cloud-provider flag is given, try to use data from cadvisor to
determine the ProviderID.

@smarterclayton ptal
@simon3z  @enoodle fyi

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2865,6 +2870,17 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
 	return nil
 }
 
+func (kl *Kubelet) updateCloudProviderFromMachineInfo(
+	node *api.Node,
+	info *cadvisorapi.MachineInfo) {
+
+	if info.CloudProvider != cadvisorapi.UnkownProvider &&
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: The typo causes me pain here.

**Comment by smarterclayton** [comment]
Looks ok to me.  Is the holdup upstream convincing vish?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go (hunk 1 above)
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: Pretty sure we'll pick up the typo fix with the imminent cadvisor bump

**Comment by liggitt** [comment]
typo fix is in https://github.com/openshift/origin/pull/8072

**Comment by deads2k** [comment]
> Looks ok to me. Is the holdup upstream convincing vish?

Brian punted it from 1.2 (not important enough I'd guess), so no one's looked at it again.  Looked reasonable to me too.  I'll update the typo when the new level merges.

**Comment by deads2k** [comment]
Updated on top of the 1.2 tag.

@smarterclayton anything else holding this up?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2865,6 +2870,17 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
...
 }
 
+func (kl *Kubelet) updateCloudProviderFromMachineInfo(
+	node *api.Node,
+	info *cadvisorapi.MachineInfo) {
+
+	if info.CloudProvider != cadvisorapi.UnknownProvider &&
+		info.CloudProvider != cadvisorapi.Baremetal {
+		node.Spec.ProviderID = strings.ToLower(string(info.CloudProvider)) +
+			":////" + string(info.InstanceID)
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: Shouldn't this be `//`?  That's what GetInstanceProviderID() returns

Comment (enoodle) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: Usually it returns <cloud-provider>://<project_name>/<avilability_zone>/<instance_id> . we dont have the project name or the availability zone so they are left empty (kubelet's cloud providers may also return them empty when they are not available)

**Comment by deads2k** [comment]
one green, one infrastructure flake.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by ncdc** [comment]
This either needs to be reverted or we need to push for https://github.com/google/cadvisor/pull/1171 to be merged ASAP so we can pull it in to origin. https://bugzilla.redhat.com/show_bug.cgi?id=1319439

**Comment by deads2k** [comment]
> This either needs to be reverted or we need to push for google/cadvisor#1171 to be merged ASAP so we can pull it in to origin. https://bugzilla.redhat.com/show_bug.cgi?id=1319439

If cadvisor actually merges that pull, I'm ok picking it down here.  Otherwise, I'd revert this.  I planned to give it until tomorrow PM.

**Pull request merged**